### ISSUE TITLE
Multiple Delegates

### ISSUE THREAD

**Issue by FintanH** [radicle-dev/heartwood#126]
As I explore the implementation of multiple delegates I wanted to track some ideas and questions down in an issue.

My approach to this is to tackle the implementation from the CLI inwards, taking `rad edit` as inspiration. I can see two ways for doing this UX-wise:
1. Similar to `rad edit` where the editor is opened and the person can add/remove delegates and save
2. Have subcommands `--add`, `--remove`, and `--list`

What would be the more clients-team-esque approach? And in the case of going down the route of 2. then what's the usual approach for doing subcommands? I'm looking at `patch` but it's not so clear to me how to generalise that to other scenarios of subcommands.

**Comment by cloudhead** [comment]
Hmmm. How about `rad delegate (add|remove|list) <peer>`?

**Comment by FintanH** [comment]
Ya that works for me! I just wasn't sure about what the general style is for the CLI :)

**Comment by cloudhead** [comment]
Yeah I think it's still being defined. Most commands don't have sub-commands. Those that do are `patch` and `issue`, and the subcommands don't use `--`.

**Comment by FintanH** [comment]
`patch` does seem to use `--` unless I'm missing something here https://github.com/radicle-dev/heartwood/blob/master/radicle-cli/src/commands/patch.rs#L27-L39

**Comment by FintanH** [comment]
Ah, `issue` is a better example https://github.com/radicle-dev/heartwood/blob/master/radicle-cli/src/commands/issue.rs#L21-L25, I'll follow that :)

**Comment by cloudhead** [comment]
> `patch` does seem to use `--` unless I'm missing something here https://github.com/radicle-dev/heartwood/blob/master/radicle-cli/src/commands/patch.rs#L27-L39

you're right, I mistaked the intended interface with the actual interface :stuck_out_tongue:, `issues` is a good example.

**Comment by FintanH** [comment]
> > `patch` does seem to use `--` unless I'm missing something here https://github.com/radicle-dev/heartwood/blob/master/radicle-cli/src/commands/patch.rs#L27-L39
> 
> you're right, I mistaked the intended interface with the actual interface stuck_out_tongue, `issues` is a good example.

I can make a PR :eyes:

**Comment by cloudhead** [comment]
> 

Yeah, we need to settle on default behavior when no sub-command is used.

Right now, `rad patch` creates a patch, and listing is via `rad patch --list`.
I think that should be changed to match `issue`:

`$ rad patch` => patch status, for now `rad patch --list` equivalent
`$ rad patch n[ew]` => create patch

I *kind of* would like to support both `--` for sub-commands as well :thinking:, ie. both of these would do the same thing:

* `rad patch --new`
* `rad patch new`

**Comment by gsaslis** [comment]
I think I am more in favour of the explicit subcommands rather than introducing subcommands as parameters. 

It seems to me the latter would make it more confusing to decide what's a subcommand vs. just a simple parameter, where there would be no consistent way to tell between the two (e.g. in something like: `rad issue edit 12345 --comment 'some change'` vs. `rad issue --edit 12345 --comment 'some change'` )

On another, more general, note: 

if we consider `patch` as a command that has several subcommands, I would expect the default invocation of `rad patch` to return with just usage instructions, explaining the different subcommands available to me. While this makes little difference in the actual functionality of the tool, I've found this makes for a smoother developer experience / onboarding onto the tool. 

The alternative of mapping `rad patch` to the invocation of e.g. `rad patch list` (or `rad patch --list`) raises the question  of why should the listing be the default and not something else. I do recognise that there is an established pattern for listing items by default e.g. on REST-like API endpoints, but I am not so sure that the same pattern is the standard in Command Line Interface (CLI) tools. 

To sum up, I think I'd expect: 

* `rad $command` => usage instructions about command
* `rad $command $subcommand` => usage instructions and options about subcommand
* `rad $command $subcommand $args` => execute $subcommand of $command with $args (e.g. `rad issue close 12345` ) 
* `rad $command $subcommand --$parameter0 $optional_param0_value --parameter1 $optional_param1_value $args` => execute $subcommand of $command with $args and, well, you get the point... :)

**Comment by cloudhead** [comment]
> The alternative of mapping `rad patch` to the invocation of e.g. `rad patch list` (or `rad patch --list`) raises the question of why should the listing be the default and not something else. I do recognise that there is an established pattern for listing items by default e.g. on REST-like API endpoints, but I am not so sure that the same pattern is the standard in Command Line Interface (CLI) tools.

This is pretty established in git already, eg. see `git branch` and `git remote`

**Comment by gsaslis** [comment]
Fair point! 

I should have just stated it's bad UX - and that even tools like `git` suffer from bad UX practices. ;)

**Comment by FintanH** [comment]
> I _kind of_ would like to support both `--` for sub-commands as well thinking, ie. both of these would do the same thing:
> 
>     * `rad patch --new`
> 
>     * `rad patch new`

Why's that? It seems like it would be confusing and also a PITA to maintain since we're doing a lot of manual work to parse args as it is.

**Comment by cloudhead** [comment]
Just because some people might be used to the `--new` style and it would avoid having to remember to switch invocation style everytime you use `rad`. But tbf I don't really care, I'm fine with just supporting the `rad patch new` syntax and seeing how that pans out.

**Comment by FintanH** [comment]
This should be "done" now with `rad delegate` and `rad id` in place -- modulo potential bugs :)
